semgrep-core/src/fixing/Autofix_printer.mli

```diff
@@ -0,0 +1,10 @@
+(* Print the given AST to a string, using the original text of nodes from the
+ * fix pattern or target file wherever possible *)
+val print_ast :
+  lang:Lang.t ->
+  metavars:Metavariable.bindings ->
```

Comment:
why do we need the metavars again given the fix_pattern_ast has already been gone though
the Metavar_replacement phase?

Reply:
We need all of these arguments for use in the external printer. The final argument to this function is the AST that we want to print. However, we also need the metavar bindings so that we can tell which nodes in the final AST came unchanged from a metavar, and the target contents so that we can lift the original text for those nodes from it.

Similarly, we need the AST for the original fix pattern, as well as that fix pattern's text, so that we can tell which nodes came from the pattern and use their original text.

I'll add a comment to this effect.

Reply:
I was also confused about this :D